When running the glibc tests in a container or chroot under a
system with a newer Linux kernel the tst-cancel4, tst-cancelx4,
tst-cancel5, and tst-cancdlx5 tests may fail to block on a
socket write because the write buffer size is too small. This
results in a test failure. The fix is to increase the write buffer
size enough to block.

commit 135529b443631f840cc66d0cc395f79c416434d9
Author: David S. Miller <davem@davemloft.net>
Date:   Tue Jul 23 11:31:39 2013 -0700

    Remove Linux kernel version ambiguity in comment added by previous commit.
    
        * tst-cancel4.c (WRITE_BUFFER_SIZE): Adjust comment.

commit e7074e4c5edb0acaa979ea08e533736f906a9d68
Author: David S. Miller <davem@davemloft.net>
Date:   Tue Jul 23 02:31:37 2013 -0700

    Increase nptl test case buffer size so we really block on current Linux kernels.
    
        * tst-cancel4.c (WRITE_BUFFER_SIZE): Increase to 16384.

Index: glibc-2.12-2-gc4ccff1/nptl/tst-cancel4.c
===================================================================
--- glibc-2.12-2-gc4ccff1.orig/nptl/tst-cancel4.c
+++ glibc-2.12-2-gc4ccff1/nptl/tst-cancel4.c
@@ -84,7 +84,30 @@ static pthread_barrier_t b2;
 # define IPC_ADDVAL 0
 #endif
 
-#define WRITE_BUFFER_SIZE 4096
+/* The WRITE_BUFFER_SIZE value needs to be choosen such that if we set
+   the socket send buffer size to '1', a write of this size on that
+   socket will block.
+
+   The Linux kernel imposes a minimum send socket buffer size which
+   has changed over the years.  As of Linux 3.10 the value is:
+
+     2 * (2048 + SKB_DATA_ALIGN(sizeof(struct sk_buff)))
+
+   which is attempting to make sure that with standard MTUs,
+   TCP can always queue up at least 2 full sized packets.
+
+   Furthermore, there is logic in the socket send paths that
+   will allow one more packet (of any size) to be queued up as
+   long as some socket buffer space remains.   Blocking only
+   occurs when we try to queue up a new packet and the send
+   buffer space has already been fully consumed.
+
+   Therefore we must set this value to the largest possible value of
+   the formula above (and since it depends upon the size of "struct
+   sk_buff", it is dependent upon machine word size etc.) plus some
+   slack space.  */
+
+#define WRITE_BUFFER_SIZE 16384
 
 /* Cleanup handling test.  */
 static int cl_called;
